src/nemos/solvers.py

```diff
@@ -0,0 +1,725 @@
+        ValueError
+            If not all arguments in args have the same sized first dimension.
+        """
+        n_points_per_arg = {leaf.shape[0] for leaf in jax.tree.leaves(args)}
+        if not len(n_points_per_arg) == 1:
+            raise ValueError("All arguments must have the same sized first dimension.")
+        N = n_points_per_arg.pop()
+
+        m = (N + self.batch_size - 1) // self.batch_size  # number of iterations
+        # m = N
+
+        xs, df_xs = state.xs, state.df_xs
+
+        def inner_loop_body(_, carry):
+            xk, key = carry
+
+            # sample mini-batch or data point
+            key, subkey = random.split(key)
+            ind = random.randint(subkey, (self.batch_size,), 0, N)
+
+            # perform a single update on the mini-batch or data point
+            xk = self._xk_update_step(
+                xk,
+                xs,
+                df_xs,
+                state.stepsize,
+                prox_lambda,
+                *tree_slice(args, ind),
+            )
+
+            return (xk, key)
+
+        xk, key = lax.fori_loop(
+            0,
+            m,
+            inner_loop_body,
+            (current_params, state.key),
+        )
+
+        # update the state
+        # storing the average over the inner loop to potentially use it in the run loop
+        state = state._replace(
+            iter_num=state.iter_num + 1,
+            key=key,
+        )
+
+        # the next anchor point is the parameters at the end of the inner loop
+        # (or the average over the inner loop)
+        return OptStep(params=xk, state=state)
+
+    @staticmethod
+    def _error(x, x_prev, stepsize):
+        """
+        Calculate the magnitude of the update relative to the parameters.
+        Used for terminating the algorithm if a certain tolerance is reached.
+
+        Params
+        ------
+        x :
+            Parameter values after the update.
+        x_prev :
+            Previous parameter values.
+
+        Returns
+        -------
+        Scaled update magnitude.
+        """
+        # stepsize is an argument to be consistent with jaxopt
+        return tree_l2_norm(tree_sub(x, x_prev)) / tree_l2_norm(x_prev)
+
+
+class SVRG(ProxSVRG):
+    """
+    SVRG solver
+
+    Equivalent to ProxSVRG with prox as the identity function and prox_lambda=None.
+
+    Attributes
+    ----------
+    fun: Callable
+        smooth function of the form ``fun(x, *args, **kwargs)``.
+    maxiter : int
+        Maximum number of epochs to run the optimization for.
+    key : jax.random.PRNGkey
+        jax PRNGKey to start with. Used for sampling random data points.
+    stepsize : float
+        Constant step size to use.
+    tol: float
+        Tolerance level for the error when comparing parameters
+        at the end of consecutive epochs to check for convergence.
+    batch_size: int
+        Number of data points to sample per inner loop iteration.
+
+    Example
+    -------
```

Comment:
same point about this not rendering correctly

Reply:
Updated.